### PR TITLE
Bind to all the SDL2 DropEvents

### PR DESCRIPTION
**Pull request by zbaylin**
This binds to all the SDL2 drop events:
- `SDL_DROPFILE`
- `SDL_DROPTEXT`
- `SDL_DROPBEGIN`
- `SDL_DROPCOMPLETE`
It also adds a convenience function in the C++ stubs to get mouse coordinates while the window isn't in focus, which is common when dragging & dropping files.

See https://wiki.libsdl.org/SDL_DropEvent for more info

### CODE REVIEW

File: src/sdl2.re
Comment (glennsl) [inline src/sdl2.re]: We can use 4.08 features here, can't we?

```suggestion
        Option.value(file, ~default=""),
```

If not, a convenience function might be nice. Putting an entire `switch` in a function argument seems a bit much at least.

**Comment by zbaylin** [comment]
Just tested this on Windows and it works there as well!

File: src/sdl2_wrapper.cpp
Comment (glennsl) [inline src/sdl2_wrapper.cpp:666]: Definitely outside the scope of this PR, but these names are terrible. To understand what this new code does I have to both scroll up here to the beginning to see their definition, and down to the end of the 200-line long function to see how it's actually used.

Also wondering if there's a reason why the switch cases aren't more self-contained. Since this is C++, not C, you should be able to define new variables and `return` instead of `break`ing, right? Or do the `CAML` macros do weird things that makes this not possible?

File: src/sdl2.re
Comment (glennsl) [reply src/sdl2.re]: `package.json` says `>= 4.6.0 && < 4.9.0`, so I guess not.

Comment (zbaylin) [reply src/sdl2.re]: Interesting I never knew about those 4.08 features. Thanks! That's an easy fix.

Comment (zbaylin) [reply src/sdl2.re]: Hmm I see that now too. I can make a convenience function for that.

Comment (glennsl) [reply src/sdl2.re]: Yeah, 4.08 adds several nice-to-have modules to `Stdlib`. It's the main reason we moved Oni2 and made all its dependencies compatible with it. But we might still want the dependencies to be more portable.

File: src/sdl2_wrapper.cpp
Comment (zbaylin) [reply src/sdl2_wrapper.cpp:666]: Definitely agree with you that it's confusing. The `CAML` macros make things harder with regards to returning, because of intricacies with the garbage collector (IIRC). Is there something that you think I should change to make it easier to understand in the mean time? We could write inline functions for each event to make things a little clearer.

File: src/sdl2.re
Comment (zbaylin) [reply src/sdl2.re]: Added an `unwrap` function in the latest commit

File: src/sdl2_wrapper.cpp
Comment (glennsl) [reply src/sdl2_wrapper.cpp:666]: The [docs](https://caml.inria.fr/pub/docs/manual-ocaml/intfc.html#ss:c-simple-gc-harmony) just say that "Every occurrence of return x must be replaced by CAMLreturn (x)" though, so I would think it should be OK to use in place of the `break`s here.

> Is there something that you think I should change to make it easier to understand in the mean time? We could write inline functions for each event to make things a little clearer.

Nah. I agree it would be cleaner, but too much out of scope for this PR I think. I mention it just as an interesting example to learn from. Perhaps it could be cleaned up later in a separate PR though. I also think giving the variables proper names and using `return` instead of `break` (if possible) would fix most of the problems and is pretty low-hanging fruit in comparison.

Comment (zbaylin) [reply src/sdl2_wrapper.cpp:666]: Sounds good - I can rename the variables to something a little more clear.

Comment (zbaylin) [reply src/sdl2_wrapper.cpp:666]: I changed the ambiguous `x, y` in the new commit and changed some other names as well. Let me know what you think!

Comment (glennsl) [reply src/sdl2_wrapper.cpp:666]: Great! I didn't even notice those were a bit ambiguous, but that definitely helps my understanding!

Comment (zbaylin) [reply src/sdl2_wrapper.cpp:666]: Sounds good. Do you want me to change those `break`s to `return`s in this PR, or do it in a separate one?

Comment (glennsl) [reply src/sdl2_wrapper.cpp:666]: That and and renaming the `v` and `vInner` variables in a separate PR would be a good idea I think. Assuming it works!

**Comment by zbaylin** [comment]
Awesome! @glennsl, do you think this is ready for merging? CI has passed. I'll start that other PR after this one is merged.

File: src/sdl2_wrapper.cpp
Comment (CrossR) [inline src/sdl2_wrapper.cpp:842]: Maybe I'm missing something here, (which is entirely possible since I've not exactly done much in this repo).

But can these 4 cases be dropped to 2 cases with an if statement at the start (like the mouse down/up events).
If I'm being dumb, just ignore/say so! I could be missing something, but right now it looks like there is two duplicated bits with just a different tag in the `caml_alloc`. Could potentially go even further and squash to one, but that seems a bit messy.

Obviously doesn't change functionality at all, just thinking about if this gets changed at any point.

Comment (zbaylin) [reply src/sdl2_wrapper.cpp:842]: You're definitely right. Not sure why I didn't think of that. Will fix that in a sec.

Comment (glennsl) [reply src/sdl2_wrapper.cpp:842]: My preference would be to factor out the common code to a function instead. That kind of special casing quickly leads to hairy spaghetti code as requirements change and the code evolves. Abstractions like this are what functions are for after all.

**Comment by glennsl** [comment]
> do you think this is ready for merging?

It looks good to me, but I don't have a lot of experience with FFI, so getting a green light from @bryphe would be a good idea IMO.

**Comment by zbaylin** [comment]
@glennsl Sounds good. I'll wait for @bryphe to chime in and then we can merge.

File: src/sdl2_wrapper.cpp
Comment (bryphe) [reply src/sdl2_wrapper.cpp:666]: Thanks for the help here! Sorry, the unclear variables were my fault. 

I'm not sure if it is possible to use `CAMLreturn` multiple times in a function - I thought it could potentially cause problems, but I guess they do use that in the OCaml runtime, ie: https://github.com/ocaml/ocaml/blob/29d5f485d810c7f88d4513fbffde7a9dcacad3f4/runtime/bigarray.c#L1013 - so it should be OK to refactor this. 

But fine to move to another PR - thank you for the help improving this, @zbaylin !

File: src/sdl2_wrapper.cpp
Comment (bryphe) [inline src/sdl2_wrapper.cpp:826]: Thanks for handling these `free`'s 👍

File: src/sdl2_wrapper.cpp
Comment (bryphe) [inline src/sdl2_wrapper.cpp:834]: Nice job constructing these objects - can be tricky to get right. Looks great!

**Review by bryphe — APPROVED** [review]
Looks good to me! Thanks for adding this, @zbaylin 🎉

**Pull request merged**
